### ISSUE TITLE
Investigate Flatpak builds

### ISSUE THREAD

**Issue by eliandoran** [TriliumNext/Notes#233]
Not sure how popular is the Flatpak build for Trilium, but it's worth investigating how to set it up.

Regarding this, note:

Apparently Zadam downgraded Electron from 28.1.0 to 25.9.8 on [Jan 4th 2024](https://github.com/zadam/trilium//commit/394530921e3f7ddd852fa4ceb1ac4585447df35e). First release to have it is 0.63.0-beta.

Found a potential reason in [v0.63.1.1-beta](https://github.com/zadam/trilium/releases/tag/v0.63.1.1-beta):

> downgrades to try to fix flatpak build

**Comment by root-hal9000** [comment]
@eliandoran I actually forgot, but I just checked and realized I use the flatpak version. I don't know much about this, but would be willing to investigate and learn here (I was doing some light reading on building for flatpaks a few months ago).

On a side note: I am on an arch based linux distribution, so I had two options for my desktop client: flatpak or the AUR. Does anyone know who happens to be the person maintaining the [AUR package](https://aur.archlinux.org/packages/trilium)? it was updated as recently as march of this year

**Comment by nebulon42** [comment]
I have a draft PR for an update to the Flatpak build over at https://github.com/TriliumNext/trilium-flathub/pull/2.
There is still a caveat in that `electron-forge` tries to make a network request which is forbidden if the app should be build on Flathub. See description in the PR. Definitely not sure, but maybe the Electron downgrade had something to do with that problem.

**Comment by pano9000** [comment]
I've been doing a tiny bit of work today on getting "local" (i.e. without Flathub "integration") Flatpak builds done via electron-forge.
Progress can be found here:
https://github.com/TriliumNext/Notes/tree/build_add-flatpak-electron

It builds locally now, but required some extra flatpak related steps, taken from here:
https://github.com/electron/forge/issues/2561#issuecomment-1037238319

will continue some testing and then adjusting of the Github CI Workflow

@nebulon42 
It appears that your effort is aimed more towards Flathub packaging/releasing — which I have even less an idea bout.
But Id personally see this one tracked in here: https://github.com/TriliumNext/Notes/issues/295

**Comment by pano9000** [comment]
so another update here:

The built Flatpak installs fine on my machine, but has two minor issues that I'd need to address still:

#### Issue 1)
1) It looks like it uses some EOL SDK version and this prints the following, despite me having installed the latest one 

`flatpak install flathub org.freedesktop.Platform/x86_64/24.08 org.freedesktop.Sdk/x86_64/24.08 org.electronjs.Electron2.BaseApp/x86_64/stable -y `
```
Info: (pinned) runtime org.freedesktop.Platform branch 19.08 is end-of-life, with reason:
   The Freedesktop SDK 19.08 runtime is no longer supported as of September 1, 2021. Please ask your application developer to migrate to a supported version
Info: applications using this runtime:
   com.github.trilium

Info: runtime org.freedesktop.Platform.GL.default branch 19.08 is end-of-life, with reason:
   The Freedesktop SDK 19.08 runtime is no longer supported as of September 1, 2021. Please ask your application developer to migrate to a supported version
Info: applications using this extension:
   com.github.trilium
```

I'll need to do some search on this, maybe it comes form the Electron2 "package"?

#### Issue 2)
I need to figure out where that id comes from: `com.github.trilium` – we likely should rename it to ...triliumnext or something like that, but I can't find that exact string in the src – I assume it gets "crafted" by electron-forge somehow → need to research, where exactly and how to fix it

**Comment by nebulon42** [comment]
Great that you work on that. Yes, Flathub builds are a goal but I'm already using a Flatpak of TrilliumNext that I have build locally with internet access enabled. You can have a look on my PR and compare with your own work given that I also use electron-forge and 24.08.

**Comment by pano9000** [comment]
@nebulon42  thanks for the reply, out of curiosity: when you say you use "electron-forge" as well – may I ask what commands you run? or are you referring to the following line in your [com.github.triliumnext.trilium.yml](https://github.com/TriliumNext/trilium-flathub/pull/2/files#diff-5a350ce8cb03d94d1e635caf5e22d94c3546320dee814cd03a6833cbf9c4e02f)
`npx electron-forge make --targets="@electron-forge/maker-zip"`

just for my understanding :-)
because what I am essentially trying to get to work is a "minimal" Flatpak build via `npx electron-forge make --targets="@electron-forge/maker-flatpak"` directly. This way we could have Flatpak builds within the existing Github wokflows.

Your approach is probably the "more correct" way to handle "proper" Flatpaks though, because this is what you would need to actually, to publish Flatpaks on Flathub as well, from what I quickly read on Flathub.

**Comment by pano9000** [comment]
another update here: 
Working on the Github workflow now – 
* The x64 version now gets build correctly via Github runners in my tests
* for arm64 I still need to get some arm64 "quirks" correct for the workflow to build (or rather aarch64 – as used by Flatpak)

**Comment by pano9000** [comment]
edit: I think there is an upstream issue with electron-forge and arm64
Flatpak seems to only support "aarch64" as architecture code, but electron-forge does not support that:

When running `npx electron-forge make --arch=aarch64`, it will throw an error 
```
Unsupported arch=aarch64 (string); must be a string matching: ia32, x64, armv7l, arm64, mips64el, universal
```

However using `npx electron-forge make --arch=arm64` fails as well, when it tries to download from the Flatpak repo.
electron-forge tries to call the following command, which fails:
`$ flatpak install --user --no-deps --arch arm64 --assumeyes runtime/org.freedesktop.Platform/arm64/24.08`

When I run the same command manually, I get some more useful error output from Flatpak

```
Error: No remote refs found for ‘runtime/org.freedesktop.Platform/arm64/24.08’
```

If I then manually adjust the architecture like below, Flatpak is happy:
`flatpak install --user --no-deps --arch aarch64 --assumeyes runtime/org.freedesktop.Platform/aarch64/24.08`


For x64, electron-forge seems to have extra logic built in to handle it correctly.
The debug output of `npx electron-forge make --arch=x64`, shows that it used x86_64 for the flatpak commands here

```flatpak update --no-deps --arch x86_64 --assumeyes org.freedesktop.Sdk 24.08```


#### Long story short:
→ I'll finish up the x64 build workflow this evening, and will then report/fix upstream, so that we can also have arm64/aarch64 builds via Github CI at some point later as well :-)

**Comment by pano9000** [comment]
Upstream PR submitted:
https://github.com/electron/forge/pull/3839

now back to finishing the x64 build workflow :-)

**Comment by nebulon42** [comment]
@pano9000 Nice progress.
Yes, I was referring to that line you mentioned. I did not execute any commands manually for building but used flatpak-builder and the file as described in the README in the PR. Needed to use the ZIP approach because flatpak-builder does the work regarding Flatpak building and then I couldn't use the maker-flatpak from electron-forge because that is meant to be used standalone.
